Implement Default for some alloc/core iterators

the8472 opened the pull request:
Add `Default` impls to the following collection iterators:

* slice::{Iter, IterMut}
* binary_heap::IntoIter
* btree::map::{Iter, IterMut, Keys, Values, Range, IntoIter, IntoKeys, IntoValues}
* btree::set::{Iter, IntoIter, Range}
* linked_list::IntoIter
* vec::IntoIter

and these adapters:

* adapters::{Chain, Cloned, Copied, Rev, Enumerate, Flatten, Fuse, Rev}

For iterators which are generic over allocators it only implements it for the global allocator because we can't conjure an allocator from nothing or would have to turn the allocator field into an `Option` just for this change.

These changes will be insta-stable.


ACP: https://github.com/rust-lang/libs-team/issues/77

the8472 committed:
Implement Default for some alloc/core iterators

This way one can `mem::take()` them out of structs or #[derive(Default)] on structs containing them.

These changes will be insta-stable.

rust-highfive commented:
r? @m-ou-se

(rust-highfive has picked a reviewer for you, use r? to override)

joshtriplett commented:
Seems reasonable to me. Since this would be insta-stable, it needs an FCP.

@rfcbot merge

scottmcm commented on library/alloc/src/vec/into_iter.rs:
```diff
@@ -315,6 +315,13 @@ impl<T, A: Allocator> FusedIterator for IntoIter<T, A> {}
 unsafe impl<T, A: Allocator> TrustedLen for IntoIter<T, A> {}
 
+#[stable(feature = "default_iters", since = "1.64.0")]
+impl<T> Default for IntoIter<T> {
```
suggestion: put a doc comment on all these impls (or the method, I forget which we do) to *say* what they do.

the8472 replied:
Done

scottmcm commented on library/core/tests/slice.rs:
```diff
@@ -224,6 +226,12 @@ fn test_iterator_count() {
+#[test]
+fn test_iterator_default() {
+    let iter: Iter<'_, u8> = Iter::default();
+    assert_eq!(iter.len(), 0);
```
For example, this could be an example in the doc comment, for users to see, instead of a normal test.

the8472 replied:
I tried but since it's generated in a macro I get this

```
4 | /// let iter: &mut[] = Default::default();
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ rustdoc does not generate documentation for macro invocations
```

Is there a way around that?

the8472 replied:
Nevermind, figured it out.

the8472 committed:
rewrite iterator `Default` tests as doctests

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7679546263?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/0373632c854ffaf45ef287bba10a7a2bcd44e593/checks/7679546263/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  |
6 | use core::slice::Iter;
  |     ^^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
error: could not compile `core` due to previous error
Build completed unsuccessfully in 0:01:28
```

</details>

WaffleLapkin commented:
> For iterators which are generic over allocators it only implements it for the global allocator because we can't conjure an allocator from nothing or would have to turn the allocator field into an Option just for this change.

I don't think that's true. Can't we just require `A: Default`? [`Global` implements `Default`](https://doc.rust-lang.org/std/alloc/struct.Global.html#impl-Default) so there shouldn't be any problems with it.

thomcc commented:
This finished FCP. Would you mind rebasing it to resolve the conflicts?

@rustbot author

the8472 commented:
Rebased.

> I don't think that's true. Can't we just require `A: Default`? [`Global` implements `Default`](https://doc.rust-lang.org/std/alloc/struct.Global.html#impl-Default) so there shouldn't be any problems with it.

Good idea. On the assumption that's a small enough change that it won't invalidate the FCP I can add that.

the8472 committed:
Support allocators in various Default for IntoIter impls

Global implements Default so we can use that as bound for all allocators

the8472 commented:
* added Allocator support for `impl Default for IntoIter` for the Vec, BtreeMap and BtreeSet IntoIters
* updated the stable annotation to use the CURRENT_RUSTC_VERSION placeholder

@rustbot ready

WaffleLapkin commented on library/alloc/src/collections/btree/map.rs:
```diff
@@ -2021,6 +2108,20 @@ impl<K, V, A: Allocator + Clone> ExactSizeIterator for IntoKeys<K, V, A> {
 impl<K, V, A: Allocator + Clone> FusedIterator for IntoKeys<K, V, A> {}
 
+#[stable(feature = "default_iters", since = "CURRENT_RUSTC_VERSION")]
+impl<K, V> Default for IntoKeys<K, V> {
```
This can also support allocator.

the8472 replied:
done

WaffleLapkin commented on library/alloc/src/collections/btree/map.rs:
```diff
@@ -2055,6 +2156,20 @@ impl<K, V, A: Allocator + Clone> ExactSizeIterator for IntoValues<K, V, A> {
 impl<K, V, A: Allocator + Clone> FusedIterator for IntoValues<K, V, A> {}
 
+#[stable(feature = "default_iters", since = "CURRENT_RUSTC_VERSION")]
+impl<K, V> Default for IntoValues<K, V> {
```
This can also support allocator.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3351295040/jobs/5552638963) [(plain)](https://github.com/rust-lang/rust/commit/ae9bdb39d86160ce69e61dcf2513ac56ab34a54f/checks/9179013301/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/alloc/src/vec/in_place_collect.rs at line 137:
 //! }
 //! vec.truncate(write_idx);
 //! ```
+use crate::alloc::Global;
 use core::alloc::{Allocator, Layout};
 use core::iter::{InPlaceIterable, SourceIter, TrustedRandomAccessNoCoerce};
 use core::mem::{self, ManuallyDrop, SizedTypeProperties};
Diff in /checkout/library/alloc/src/vec/in_place_collect.rs at line 143:
 use core::ptr::{self, NonNull};
 
 
 use super::{InPlaceDrop, InPlaceDstBufDrop, SpecFromIter, SpecFromIterNested, Vec};
Diff in /checkout/library/alloc/src/vec/in_place_collect.rs at line 209:
 
 
         // launder the pointer through the allocator if the alignedment changed
         // ideally this does not cause any data movement
-        if mem::align_of::<T>() != mem::align_of::<<<I as SourceIter>::Source as AsVecIntoIter>::Item>() {
+        if mem::align_of::<T>()
+            != mem::align_of::<<<I as SourceIter>::Source as AsVecIntoIter>::Item>()
             let alloc = Global;
             unsafe {
-                let reallocated = alloc.shrink(
-                let reallocated = alloc.shrink(
-                    NonNull::new_unchecked(dst_buf as *mut u8),
-                    Layout::array::<<<I as SourceIter>::Source as AsVecIntoIter>::Item>(cap).unwrap(),
-                    Layout::array::<T>(cap).unwrap()
-                ).expect("failed to realloc vec");
+                let reallocated = alloc
+                    .shrink(
+                        NonNull::new_unchecked(dst_buf as *mut u8),
+                        Layout::array::<<<I as SourceIter>::Source as AsVecIntoIter>::Item>(cap)
+                            .unwrap(),
+                        Layout::array::<T>(cap).unwrap(),
+                    )
+                    .expect("failed to realloc vec");
                 dst_buf = reallocated.as_ptr() as *mut T;
         }
         }
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/alloc/src/vec/drain_filter.rs" "/checkout/library/alloc/src/vec/spec_from_elem.rs" "/checkout/library/alloc/src/vec/in_place_drop.rs" "/checkout/library/alloc/src/vec/in_place_collect.rs" "/checkout/library/alloc/src/vec/spec_extend.rs" "/checkout/library/alloc/src/vec/spec_from_iter.rs" "/checkout/library/alloc/src/vec/mod.rs" "/checkout/library/alloc/src/vec/partial_eq.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #106833) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #107634) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

scottmcm reviewed (approved):
I did a quick scan and I think all these are fine, but for future things I wanted to drop a note to be careful around `ManuallyDrop` vs `MaybeUninit` differences in various iterators.  `array::IntoIter` could *currently* have an empty `Default`, but it's possible that it shouldn't actually allow that at all, since [arguably](https://github.com/rust-lang/rust/issues/91583#issuecomment-1414745279) it should be implemented with `ManuallyDrop`, not `MaybeUninit`.

the8472 commented:
Can the FCP be taken as libs-api approval and yours covering the implementation side?

scottmcm commented:
I just did a quick scan for that, since I'd been in "requested changes".  I'll take the review, though, and can probably sign-off soon.

r? @scottmcm

scottmcm commented on library/core/src/iter/adapters/fuse.rs:
```diff
@@ -181,6 +181,21 @@ where
 }
 
+#[stable(feature = "default_iters", since = "CURRENT_RUSTC_VERSION")]
+impl<I: Default> Default for Fuse<I> {
```
Pondering: I noticed that the other adapters require `I: Iterator` in addition to just `I: Default`.

Should this one require that too?

the8472 replied:
Not sure, the bounds on the adapters might have been unintentional due to copypasta. Removing them would also be an option.

scottmcm replied:
Hello @rust-lang/libs-api folks!  This PR, as currently written, allows `<Fuse<u32>>::default()` but not `<Rev<u32>>::default()`.  That feels unintentional, so nominating for a decision on what the intent was here -- notably there wasn't an API summary in the FCP so it's hard to know for sure.

Do you want:
- The current state of the PR
- The adapters to be `Default` only when wrapping iterators
- The adapters to be `Default` when wrapping any inner type that's `Default`

m-ou-se replied:
I think the third option would be best. Relaxing the bounds as much as possible seems consistent with how we don't have bounds on the types (only on the methods), and it would play a bit better with `#[derive(Default)]` when putting a `Adapter<T>` in a struct.

Nobody in today's libs-api meeting was opposed to the third option, but @dtolnay and @BurntSushi weren't available today. (Ping! Do you have any strong opinions on this?)

the8472 replied:
latest commit removes the bounds

the8472 committed:
relax bounds on iterator adapter Default impls

scottmcm commented on library/alloc/src/collections/btree/map.rs:
```diff
@@ -421,6 +449,23 @@ impl<K: Debug, V: Debug, A: Allocator + Clone> Debug for IntoIter<K, V, A> {
+#[stable(feature = "default_iters", since = "CURRENT_RUSTC_VERSION")]
+impl<K, V, A> Default for IntoIter<K, V, A>
+where
+    A: Allocator + Default + Clone,
```
Hmm, I would have guessed just
```suggestion
    A: Default,
```
but it's unstable and we can weaken it later, so meh.

scottmcm commented on library/alloc/src/collections/btree/set.rs:
````diff
@@ -1560,6 +1575,23 @@ impl<T, A: Allocator + Clone> ExactSizeIterator for IntoIter<T, A> {
+    ///
+    /// ```
+    /// # use std::collections::btree_set;
+    /// let iter: btree_set::IntoIter<u8> = Default::default();
````
Thank you for all these tests!

scottmcm commented on library/alloc/src/vec/into_iter.rs:
```diff
@@ -347,6 +347,24 @@ impl<T, A: Allocator> FusedIterator for IntoIter<T, A> {}
+#[stable(feature = "default_iters", since = "CURRENT_RUSTC_VERSION")]
+impl<T, A> Default for IntoIter<T, A>
+where
+    A: Allocator + Default,
```
Hmm, not `Clone` for this one?

(But I guess again it's a fine "not this PR")

scottmcm commented:
@bors r+

bors commented:
:pushpin: Commit e44836faf6ff114805062f723b349d6e10bf86b6 has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm e44836faf6ff114805062f723b349d6e10bf86b6 -->
<!-- homu: {"type":"Approved","sha":"e44836faf6ff114805062f723b349d6e10bf86b6","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e44836faf6ff114805062f723b349d6e10bf86b6 with merge cfbe901999ae73e9ebd8a7ea68ba8381a91674b0...
<!-- homu: {"type":"BuildStarted","head_sha":"e44836faf6ff114805062f723b349d6e10bf86b6","merge_sha":"cfbe901999ae73e9ebd8a7ea68ba8381a91674b0"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4517766498/jobs/7957304898)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4517766498/jobs/7957304898","builder_name":"checks-actions"} -->

scottmcm commented:
@bors retry #108227

bors commented:
:hourglass: Testing commit e44836faf6ff114805062f723b349d6e10bf86b6 with merge 9fa6b3c15758e85657d5be051cfa57022a8bbe57...
<!-- homu: {"type":"BuildStarted","head_sha":"e44836faf6ff114805062f723b349d6e10bf86b6","merge_sha":"9fa6b3c15758e85657d5be051cfa57022a8bbe57"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4517914543/jobs/7958330755)
Approved by: scottmcm
Pushing 9fa6b3c15758e85657d5be051cfa57022a8bbe57 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"scottmcm","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4517914543/jobs/7958330755"},"merge_sha":"9fa6b3c15758e85657d5be051cfa57022a8bbe57"} -->

rust-timer commented:
Finished benchmarking commit (9fa6b3c15758e85657d5be051cfa57022a8bbe57): [comparison URL](https://perf.rust-lang.org/compare.html?start=24a69af213f4bf23014dc91903e7a934c88f6b9e&end=9fa6b3c15758e85657d5be051cfa57022a8bbe57&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=24a69af213f4bf23014dc91903e7a934c88f6b9e&end=9fa6b3c15758e85657d5be051cfa57022a8bbe57&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.7%, 0.7%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.3%, 0.6%]   | 9     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.4%, -0.3%] | 6     |
| All ❌✅ (primary)                 | 0.7%  | [0.7%, 0.7%]   | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=24a69af213f4bf23014dc91903e7a934c88f6b9e&end=9fa6b3c15758e85657d5be051cfa57022a8bbe57&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.9% | [-3.5%, -0.4%] | 14    |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=24a69af213f4bf23014dc91903e7a934c88f6b9e&end=9fa6b3c15758e85657d5be051cfa57022a8bbe57&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.7%, 0.7%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.2%, -1.2%] | 1     |
| All ❌✅ (primary)                 | 0.7%  | [0.7%, 0.7%]   | 1     |
</details>

rust-log-analyzer commented:
The job **`dist-i686-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4517766498/jobs/7957189717) [(plain)](https://github.com/rust-lang-ci/rust/commit/cfbe901999ae73e9ebd8a7ea68ba8381a91674b0/checks/12270374885/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[2588/3021] Linking CXX static library lib\libLLVMCodeGen.a
[2589/3021] Linking CXX static library lib\libLLVMCoroutines.a
[2590/3021] Linking CXX static library lib\libLLVMAsmPrinter.a
FAILED: lib/libLLVMAsmPrinter.a 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E rm -f lib\libLLVMAsmPrinter.a && C:\a\rust\rust\mingw32\bin\ar.exe qc lib\libLLVMAsmPrinter.a  lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/AccelTable.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/AddressPool.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/AIXException.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/ARMException.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/AsmPrinter.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/AsmPrinterDwarf.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/AsmPrinterInlineAsm.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DbgEntityHistoryCalculator.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DebugHandlerBase.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DebugLocStream.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DIE.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DIEHash.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfCFIException.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfCompileUnit.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfDebug.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfExpression.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfFile.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfStringPool.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/DwarfUnit.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/EHStreamer.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/ErlangGCPrinter.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/OcamlGCPrinter.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/PseudoProbePrinter.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/WinCFGuard.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/WinException.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/CodeViewDebug.cpp.obj lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/WasmException.cpp.obj && C:\a\rust\rust\mingw32\bin\ranlib.exe lib\libLLVMAsmPrinter.a && cd ."
C:\a\rust\rust\mingw32\bin\ar.exe: could not create temporary file whilst writing archive: no more archived files
[2591/3021] Building CXX object tools/bugpoint/CMakeFiles/bugpoint.dir/ExecutionDriver.cpp.obj
[2592/3021] Linking CXX static library lib\libLLVMSelectionDAG.a
[2593/3021] Building CXX object tools/bugpoint/CMakeFiles/bugpoint.dir/OptimizerDriver.cpp.obj
[2594/3021] Building CXX object tools/bugpoint/CMakeFiles/bugpoint.dir/ExtractFunction.cpp.obj
[2594/3021] Building CXX object tools/bugpoint/CMakeFiles/bugpoint.dir/ExtractFunction.cpp.obj
[2595/3021] Building CXX object tools/bugpoint/CMakeFiles/bugpoint.dir/CrashDebugger.cpp.obj
[2596/3021] Building CXX object tools/bugpoint/CMakeFiles/bugpoint.dir/ToolRunner.cpp.obj
[2597/3021] Linking CXX executable bin\llvm-profdata.exe
ninja: build stopped: subcommand failed.
command did not execute successfully, got: exit code: 1


build script failed, must exit now', C:\Users\runneradmin\.cargo\registry\src\index.crates.io-1cd66030c949c28d\cmake-0.1.48\src\lib.rs:975:5
 finished in 300.488 seconds
Build completed unsuccessfully in 0:07:58
```

</details>

the8472 commented:
No, I don't recall intentionally excluding them.

zachs18 commented:
(For posterity: `Default for btree_map::{ValuesMut, RangeMut}` and `binary_heap::Iter` were added in https://github.com/rust-lang/rust/pull/128261)
